### PR TITLE
support partition read with partition pruning 

### PR DESCRIPTION
**Pull request by zhexuany**
To be continued.

### CODE REVIEW

**Comment by zhexuany** [comment]
It is not finished yet but can read date from tikv with partition table enabled. Also, parition pruning is working but need more attention to some corner case.

File: core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala [hunk 1]
```diff
@@ -33,6 +33,73 @@ class IssueTestSuite extends BaseTiSparkSuite {
...
+    // extract is not supported
+    // https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_extract
+    assert(spark.sql("SELECT floor(-0.1)").count() == 1)
+    assert(spark.sql("SELECT hour('2009-07-30 12:58:59')").count() == 1)
+    // microsecond is not supported
+    assert(spark.sql("SELECT minute('2009-07-30 12:58:59')").count() == 1)
+    assert(spark.sql("SELECT 2 mod 1.8").count() == 1)
+    assert(spark.sql("SELECT month('2016-07-30')").count() == 1)
+    assert(spark.sql("SELECT quarter('2016-08-31')").count() == 1)
+    assert(spark.sql("SELECT second('2009-07-30 12:58:59')").count() == 1)
```
Comment (birdstorm) [inline core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala]: do we need to check if answer is correct?

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/PartExprFunction.scala [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+  @transient private lazy val c = {
+    val c = Calendar.getInstance(DateTimeUtils.getTimeZone("UTC"))
+    c.setFirstDayOfWeek(Calendar.MONDAY)
+    c.setMinimalDaysInFirstWeek(4)
+    c
+  }
+
+  private def adjustCalendarByMode(mode: Int): Unit =
+    mode match {
```
Comment (birdstorm) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/PartExprFunction.scala]: I think we can add a link to mysql doc for these modes.

File: pom.xml [hunk 3]
```diff
@@ -172,6 +172,11 @@
             </plugins>
         </pluginManagement>
         <plugins>
+          <plugin>
+            <groupId>org.ensime.maven.plugins</groupId>
+            <artifactId>ensime-maven</artifactId>
```
Comment (birdstorm) [inline pom.xml]: are they necessary?

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiPartitionInfo.java [hunk 4]
```diff
@@ -45,13 +46,41 @@ public TiPartitionInfo(
...
 
+  private PartitionType toPartTp(long type) {
+    if (type == 0L) {
+      return PartitionType.RangePartition;
+    } else if (type == 1L) {
+      return PartitionType.HashPartition;
+    } else if (type == 2L) {
+      return PartitionType.ListPartition;
+    }
+    // only allow range, hash, range partition.
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/meta/TiPartitionInfo.java]: list?

File: tikv-client/src/main/java/com/pingcap/tikv/types/SetType.java [hunk 5]
```diff
@@ -49,7 +49,7 @@ protected Object decodeNotNull(int flag, CodecDataInput cdi) {
   /** {@inheritDoc} */
   @Override
   protected void encodeKey(CodecDataOutput cdo, Object value) {
-    throw new UnsupportedTypeException("Set type not supported");
+    throw new UnsupportedTypeException("Set type notTiPartitionInfo.java supported");
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/types/SetType.java]: is this a typo?

Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/types/SetType.java]: Yes. It is a typo.

File: .gitignore [hunk 6]
```diff
@@ -71,3 +71,6 @@ python/pytispark.egg-info/
 
 # ignore spark warehouse
 spark-warehouse
+
+.ensime
+.ensime_cache
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply .gitignore]: It is for emacs scala config.

**Comment by ilovesoup** [comment]
> To be continued.

What else you need to do? Would you specify it instead of leaving TBC.

File: .gitignore (hunk 6 above)
[the remark this answers is not in the document]
Comment (ilovesoup) [reply .gitignore]: I think it would be fine to leave those here.

File: core/src/main/scala/com/pingcap/tispark/PartitionExpr.scala [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+case class PartitionExpr(partitionExprs: List[Expression],
+                         locateExprs: List[Expression],
+                         partitionExpr: Expression,
+                         col: String) {
+  val ranges = partitionExprs
+  val partExpr = partitionExpr
+  val upperBound = locateExprs
+  val columnName = col
+
+  private def buildExprRange(e: Expression): Range[AsOrdered[Long]] = {
```
Comment (ilovesoup) [inline core/src/main/scala/com/pingcap/tispark/PartitionExpr.scala]: Why this cannot be in with ScanAnalyzer class? This should be with similar logic.

File: pom.xml (hunk 3 above)
Comment (ilovesoup) [reply pom.xml]: Maybe this should move to maven profile.

File: core/src/main/scala/com/pingcap/tispark/PartitionExpr.scala [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        if (!left.isInstanceOf[Literal] && right.isInstanceOf[Literal]) {
+          range = Range.open[AsOrdered[Long]](
+            AsOrdered(right.eval().asInstanceOf[Number].longValue()),
+            AsOrdered(Long.MaxValue)
+          )
+        }
+        if (!right.isInstanceOf[Literal] && left.isInstanceOf[Literal]) {
+          range = Range.open[AsOrdered[Long]](
+            AsOrdered(Long.MinValue),
+            AsOrdered(left.eval().asInstanceOf[Number].longValue())
```
Comment (ilovesoup) [inline core/src/main/scala/com/pingcap/tispark/PartitionExpr.scala]: Why cast to integer here?

File: core/src/main/scala/com/pingcap/tispark/PartitionExpr.scala [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+          )
+        }
+        range
+      }
+      case LessThan(left, right) => {
```
Comment (ilovesoup) [inline core/src/main/scala/com/pingcap/tispark/PartitionExpr.scala]: Do you have thorough test on all those cases?

File: core/src/main/scala/org/apache/spark/sql/TiSparkFunctionRegistry.scala [hunk 10]
```diff
@@ -0,0 +1,114 @@
+package org.apache.spark.sql
```
Comment (ilovesoup) [inline core/src/main/scala/org/apache/spark/sql/TiSparkFunctionRegistry.scala]: Copyright.

File: core/src/main/scala/org/apache/spark/sql/TiSparkFunctionRegistry.scala [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    // misc non-aggregate functions
+    expression[Microsecond]("microsecond"),
+    expression[WeekDay]("weekday"),
+    expression[ToSeconds]("to_seconds"),
+    expression[ToDays]("to_days"),
+    expression[TimeToSec]("time_to_sec"),
+    expression[YearWeek]("yearweek")
+    // TODO: this need import antler 4 into our code base.
+    // the behavior is like the following:
+    // select extract year from '2019-01-10' which is just
```
Comment (ilovesoup) [inline core/src/main/scala/org/apache/spark/sql/TiSparkFunctionRegistry.scala]: Is this syntax supported in create table? If so, you do need our own antlr parser. If it only appears in select, why do we need this?

File: core/src/main/scala/org/apache/spark/sql/TiSparkFunctionRegistry.scala [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+  )
+
+  /** See usage above. */
+  private def expression[T <: Expression](
+    name: String
+  )(implicit tag: ClassTag[T]): (String, (ExpressionInfo, FunctionBuilder)) = {
+
+    // For `RuntimeReplaceable`, skip the constructor with most arguments, which is the main
+    // constructor and contains non-parameter `child` and should not be used as function builder.
+    val constructors = if (classOf[RuntimeReplaceable].isAssignableFrom(tag.runtimeClass)) {
```
Comment (ilovesoup) [inline core/src/main/scala/org/apache/spark/sql/TiSparkFunctionRegistry.scala]: If this logic comes from Spark, please add comments refers to spark. When we upgrade, we need to check those logic in case bug fix ignored.

**Comment by ilovesoup** [comment]
You need more test before merge. Add test for different types of partition columns, filter ranges and etc.

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/PartExprFunction.scala [hunk 13]
```diff
@@ -0,0 +1,340 @@
+package org.apache.spark.sql.catalyst.expressions
```
Comment (ilovesoup) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/PartExprFunction.scala]: copyright
